lib/Lower/Bridge.cpp

```diff
@@ -373,15 +373,22 @@ class FirConverter : public Fortran::lower::AbstractConverter {
       std::string &name, const Fortran::semantics::Symbol *&symbol) {
     setCurrentPosition(stmt.source);
     auto &n = stmt.statement.v;
-    name = n.ToString();
     symbol = n.symbol;
+    assert(symbol && "Name resolution failure");
+    name = mangleName(*symbol);
   }
   void
   genFIR(const Fortran::parser::Statement<Fortran::parser::EndMpSubprogramStmt>
              &stmt,
-         std::string &, const Fortran::semantics::Symbol *&) {
-    genFIRProcedureExit(
-        static_cast<const Fortran::parser::MpSubprogramStmt *>(nullptr));
+         std::string &, const Fortran::semantics::Symbol *&symbol) {
+    setCurrentPosition(stmt.source);
+    assert(symbol);
+    if (symbol->get<Fortran::semantics::SubprogramDetails>().isFunction()) {
+      genFIRFunctionReturn(*symbol);
```

Comment:
Can this part be merged with genFIRProcedureExit? (Don't all procedures require (semi) common exit handling code?

Reply:
That's cleaner, good point. I have completely reworked all this to only keep  `genFIRProcedureExit` and `genFIRProgramExit`.